src/modules/musig/session_impl.h

```diff
@@ -0,0 +1,717 @@
+    /* TODO Cache mu */
+    secp256k1_musig_keyaggcoef(&mu, &cache_i, &pk.x);
+    secp256k1_scalar_mul(&sk, &sk, &mu);
+
+    if (!secp256k1_musig_session_load(ctx, &session_i, session)) {
+        secp256k1_musig_partial_sign_clear(&sk, k);
+        return 0;
+    }
+
+    if (session_i.fin_nonce_parity) {
+        secp256k1_scalar_negate(&k[0], &k[0]);
+        secp256k1_scalar_negate(&k[1], &k[1]);
+    }
+
+    /* Sign */
+    secp256k1_scalar_mul(&s, &session_i.challenge, &sk);
+    secp256k1_scalar_mul(&k[1], &session_i.noncecoef, &k[1]);
+    secp256k1_scalar_add(&k[0], &k[0], &k[1]);
+    secp256k1_scalar_add(&s, &s, &k[0]);
+    secp256k1_musig_partial_sig_save(partial_sig, &s);
+    secp256k1_musig_partial_sign_clear(&sk, k);
+    return 1;
+}
+
+int secp256k1_musig_partial_sig_verify(const secp256k1_context* ctx, const secp256k1_musig_partial_sig *partial_sig, const secp256k1_musig_pubnonce *pubnonce, const secp256k1_xonly_pubkey *pubkey, const secp256k1_musig_keyagg_cache *keyagg_cache, const secp256k1_musig_session *session) {
+    secp256k1_keyagg_cache_internal cache_i;
+    secp256k1_musig_session_internal session_i;
+    secp256k1_scalar mu, e, s;
+    secp256k1_gej pkj;
+    secp256k1_ge nonce_pt[2];
+    secp256k1_gej rj;
+    secp256k1_gej tmp;
+    secp256k1_ge pkp;
+
+    VERIFY_CHECK(ctx != NULL);
+    ARG_CHECK(partial_sig != NULL);
+    ARG_CHECK(pubnonce != NULL);
+    ARG_CHECK(pubkey != NULL);
+    ARG_CHECK(keyagg_cache != NULL);
+    ARG_CHECK(session != NULL);
+
+    if (!secp256k1_musig_session_load(ctx, &session_i, session)) {
+        return 0;
+    }
+
+    /* Compute "effective" nonce rj = aggnonce[0] + b*aggnonce[1] */
+    /* TODO: use multiexp to compute -s*G + e*pubkey + aggnonce[0] + b*aggnonce[1] */
+    if (!secp256k1_musig_pubnonce_load(ctx, nonce_pt, pubnonce)) {
+        return 0;
+    }
+    secp256k1_gej_set_ge(&rj, &nonce_pt[1]);
+    secp256k1_ecmult(&rj, &rj, &session_i.noncecoef, NULL);
+    secp256k1_gej_add_ge_var(&rj, &rj, &nonce_pt[0], NULL);
+
+    if (!secp256k1_xonly_pubkey_load(ctx, &pkp, pubkey)) {
+        return 0;
+    }
+    if (!secp256k1_keyagg_cache_load(ctx, &cache_i, keyagg_cache)) {
+        return 0;
+    }
+    /* Multiplying the challenge by the KeyAgg coefficient is equivalent
+     * to multiplying the signer's public key by the coefficient, except
+     * much easier to do. */
+    secp256k1_musig_keyaggcoef(&mu, &cache_i, &pkp.x);
+    secp256k1_scalar_mul(&e, &session_i.challenge, &mu);
+
+    /* If the MuSig-aggregate point has an odd Y coordinate, the signers will
+     * sign for the negation of their individual xonly public key. If the
+     * aggregate key is untweaked, then internal_key_parity is 0, so `e` is
+     * negated exactly when the aggregate key parity is odd. If the aggregate
+     * key is tweaked, then negation happens when the aggregate key has an odd Y
+     * coordinate XOR the internal key has an odd Y coordinate.*/
```

Comment:
Suggestion:

```c
    /* e needs to be negated by the verifier if and only if sk needs to be negated
     * by the signer (see the explanation in musig_partial_sign above). Unlike the
     * the signing case, we don't need to check the parity of the signer's public key
     * because it must be even (it was an xonly_pubkey to begin with).
     */
```

Reply:
Nice, I added this. It may make sense to move these explanations to the specification eventually.